csharp/src/SeedLang/X/BaseParser.cs

```diff
@@ -39,14 +44,22 @@ internal abstract class BaseParser {
     // Parses source code into an AST tree based on the parse rule. The concrete ANTLR4 lexer and
     // parser are created by the derived class. The out node is set to null if the given source code
     // is not valid.
-    internal bool TryParse(string source, string module, ParseRule rule,
-                           DiagnosticCollection collection, out AstNode node) {
+    internal bool Parse(string source, string module, ParseRule rule,
+                        DiagnosticCollection collection, out AstNode node,
+                        out IReadOnlyList<SyntaxToken> tokens) {
       int diagnosticCount = collection.Diagnostics.Count;
       Parser parser = SetupParser(source, module, collection);
       ParserRuleContext context = GetContext(parser, rule);
-      var visitor = MakeVisitor();
-      node = visitor.Visit(context);
+      var tokenList = new List<SyntaxToken>();
+      tokens = tokenList;
+      AbstractParseTreeVisitor<AstNode> visitor = MakeVisitor(tokenList);
+      try {
+        node = visitor.Visit(context);
+      } catch (ParseException) {
+        node = null;
```

Comment:
The intention of using a new ParseException is to terminate the parsing process when some AST nodes cannot be generated. But the errors need not be collected, because they are already collected by the SyntaxErrorStrategy class.

It might be a little confused, because the ParseException is duplicated with Diagnostic Information. So I tried another solution that is removing the ParseException, and carefully returns null when any AST node cannot be generated. It works well now.

We may need other Exception classes when I start to implement byte code compiler and VM. I will think about if DiagnosticException can fit into that situation, or if we need introduce other lightweight mechanism.